src/VecSim/algorithms/hnsw/hnsw_multi.h

```diff
@@ -42,8 +42,9 @@ class HNSWIndex_Multi : public HNSWIndex<DataType, DistType> {
     inline void resizeLabelLookup(size_t new_max_elements) override;
 
     // Return all the labels in the index - this should be used for computing the number of distinct
-    // labels in a tiered index, and caller should hold the index data guard.
+    // labels in a tiered index.
     inline vecsim_stl::set<labelType> getLabelsSet() const override {
+        std::shared_lock<std::shared_mutex> index_data_lock(this->indexDataGuard);
```

Comment:
consider using the api lockSharedIndexDataGuard

Reply:
Though we are using that api in many places, I believe it is a better practice to acquire locks like this where possible